Add automation support via WebDriver

JuhaVainio opened the pull request:
This PR integrates with the automation concepts defined in
https://w3c.github.io/device-posture/#automation to allow changing device posture through the WebDriver extension
commands defined there.

Set device posture WebDriver extension will enable overriding posture defined by hardware.

Clear device posture WebDriver extension removes device posture override and returns device posture control back to hardware.

Fixes #1_4_0
(Maybe I won't put the actual number here, but then to official)

JuhaVainio committed:
Add Automation chapter to specification

JuhaVainio commented:
@rakuco Are you able to access this?

JuhaVainio committed:
Create referenece out of WEBDRIVER2-text

JuhaVainio committed:
Add parameter name to code step

rakuco commented:
> @rakuco Are you able to access this?

Yep, just haven't had time to review this yet!

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
         </section>
       </section>
     </section>
+    <section id="automation" class="informative">
```
If you say a section is informative, it is non-normative, which means it's not specifying anything (which is why the Examples section below was marked as informative).

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/bae1a77ba70c568c1f67c9b6c3470427c90c73ee

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+      <p>
+        The Device Posture API pose a challenge to test authors, as fully
+        exercising interface requires physical hardware devices. To address this
+        challenge this document define [[WEBDRIVER2]]
```
```suggestion
        challenge this document defines a [[WEBDRIVER2]]
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/e8b4c90bfb60199245dae6f40807bf2e6fabd249

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+        exercising interface requires physical hardware devices. To address this
+        challenge this document define [[WEBDRIVER2]]
+        <a href="https://w3c.github.io/webdriver/#dfn-extension-commands">
+        extension command</a> that allow controlling device posture like in real
```
```suggestion
        extension command</a> that allows users to control the reported device posture and simulate a real
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/0a1ada52a79333974a0e29c1ec2eae2d52213d9a

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+            POST
+          </td>
+          <td>
+            window/deviceposture
```
Why window?

JuhaVainio replied:
Yep maybe it should be removed.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/fd9da15cd0359f6c4b4ae1be9c5983c6f4c7ca1b

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+      </table>
+      <p>
+        This <a href="https://w3c.github.io/webdriver/#dfn-extension-commands">
+        extension command</a> changes device posture to certain
```
```suggestion
        extension command</a> changes device posture to a specific
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/e8b4c90bfb60199245dae6f40807bf2e6fabd249

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+        remote end steps</a> are:
+      </p>
+      <ol class="algorithm">
+        <li>If |posture| is not {{DevicePostureType}}, return <a
```
The algorithm needs to be as specific as possible. What does "be {{DevicePostureType}}" mean? Are you talking about types? Values?

JuhaVainio replied:
One way of writing could be 
`If |posture| is an inappropriate <a>posture</a>,`
then `<a>posture</a>` would point to https://www.w3.org/TR/device-posture/#dfn-posture
But somehow I like more when pseudo code refers to actual code (DevicePostureType enum).
Do not know if this is even bad engineering english
`If |posture| is not valid enumeration of {{DevicePostureType}},`
On the other hand I would not like that |posture| would be enum, but string as in |type| in generic sensor. I do not know if that is even possible in this interface.

rakuco replied:
Think of it like this: at this point all you have is a JS string object called "posture". The entire DevicePostureType enum is a WebIDL construct that doesn't exist in JavaScript, as it's just used when converting to and from JS and C++ to validate strings.

One thing you could try to do is use [WebIDL's own terminology](https://webidl.spec.whatwg.org/#js-enumeration) when parsing JS enum values and say "one of DevicePostureType's enumeration values", but I'm not sure how usual that is.

The other option is to validate the value by hand: check each value, and throw an error if it does not match anything you expect.

JuhaVainio replied:
Last one sounds simplest, which I try to write in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/55ec58ecbf3e4ada4bacc77d21d57ca580aa855a and code in chromedriver/window_commands.cc @ [DevicePosture_WebDriver](https://github.com/JuhaVainio/applications.browsers.thirdparty.chromium.chromium/commits/DevicePosture_WebDriver/)-branch.

I write more about actual parameter handling under this [comment](https://github.com/JuhaVainio/W3C-device-posture/pull/1/files/cba8388d4a839f2cf98da8ec13e5518a17409928#r1507285805).

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+        The Device Posture API pose a challenge to test authors, as fully
+        exercising interface requires physical hardware devices. To address this
+        challenge this document define [[WEBDRIVER2]]
+        <a href="https://w3c.github.io/webdriver/#dfn-extension-commands">
```
You're not supposed to link directly to the concepts you use. Look at the rest of the document and at the ReSpec documentation -- you need to use right notation (e.g. `[=foo=]` or `{{Bar}}`) to refer to the concepts, and ReSpec will in turn add the right links, bibliographic entries and references for you.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/e853ef9c28b8e5e3f558e7e1a12a1fee85e76acc

rakuco commented on index.html:
```diff
@@ -618,6 +618,93 @@ <h4>
+            href="https://w3c.github.io/webdriver/#dfn-invalid-argument">invalid
+            argument</a>.
+        </li>
+        <li>Set device posture as |posture|.
```
This is the most important part of the algorithm, and it is too vague: what does "set device posture" mean? How is this going to impact the rest of the algorithms so that this triggers a "change" event?

JuhaVainio replied:
I think this step should somehow reference to "[8.2 Device Posture change](https://www.w3.org/TR/device-posture/#device-posture-change)" as it seems to have all steps.

JuhaVainio replied:
I tried to rewrite the whole steps-part in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/22aab929dea40f644b329d846972df0e4e004a62

JuhaVainio committed:
Do not specify Automation-chapter as informative

Also id-field was removed as almost all chapters do not use it.

JuhaVainio committed:
Fix language errors

JuhaVainio committed:
Use ReSpec format for references, not web links

JuhaVainio committed:
Reword the purpose of Automation for Device Posture API

JuhaVainio committed:
Remove "window" word from set device posture URI Template

JuhaVainio committed:
Use ReSpec format for references, not web links

JuhaVainio committed:
Use ReSpec format for references, not web links

Remaining links removed.

rakuco commented on index.html:
```diff
@@ -618,6 +618,85 @@ <h4>
+            POST
+          </td>
+          <td>
+            deviceposture
```
Please take a look at the WebDriver spec and get more familiar with the model. "deviceposture" is not a valid URI template: it doesn't start with "/" and it's not a subpath of "/session/{session id}" like pretty much every command. See the other specs we've been working on for the kind of URI template being used.

If you just say "/deviceposture", your ChromeDriver implementation wouldn't even be able to retrieve a corresponding session.

JuhaVainio replied:
Yep in code it was /session/{session id}/deviceposture, I write it that way also in this spec

JuhaVainio replied:
Missing text added in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/d41ab3585ed56ea9826b7df8598cd60c8a5d1067

rakuco commented on index.html:
```diff
@@ -618,6 +618,85 @@ <h4>
+        The [=remote end steps=] are:
+      </p>
+      <ol class="algorithm" data-cite="WEBDRIVER2">
+        <li>If |posture| is not {{DevicePostureType}}, return [=error=] with
```
See https://w3c.github.io/webdriver/#processing-model and what the Generic Sensor spec and others do. You need to fetch this variable from somewhere; at this point, your remote end steps are given "`session`, `URL variables`, and `parameters`" (step 1.6.6).

JuhaVainio replied:
I have all along wondered should the URI be "/session/{session id}/deviceposture" with parameters or "/session/{session id}/deviceposture/{posture}". Latter would be simpler , which is always better.

But I have wondered if single parameter ("posture") is enough for this service. This is little bit connected to [Emulation.setDeviceMetricsOverride()](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/devtools/protocol/emulation_handler.cc?q=EmulationHandler::SetDeviceMetricsOverride%20file:.cc&ss=chromium%2Fchromium%2Fsrc) as it is not happy if only posture variable is provided. This can be seen in chromedriver/window_commands.cc @ [DevicePosture_WebDriver](https://github.com/JuhaVainio/applications.browsers.thirdparty.chromium.chromium/commits/DevicePosture_WebDriver/)-branch as I have to provide also width, hight, etc.

Maybe this should be first solved before going too deep in style of text.

rakuco reviewed (commented):
While here: please don't say "chapter". It doesn't have a valid meaning in this context, and the usual word is "section".

JuhaVainio committed:
Add missing "/session/{session id}/" to URI template

JuhaVainio committed:
Define clearly posture-variable possible values

JuhaVainio committed:
Rewrite "Set device posture" steps

It was not mentioned where posture-variable gets it value.

Clarified how posture information in document is updated.

Added firing an event.

rakuco commented on index.html:
```diff
@@ -618,7 +618,7 @@ <h4>
       </section>
     </section>
-    <section>
+    <section data-link-for="DevicePostureType">
```
Instead of using `data-link-for`, you can use `{{DevicePostureType/continuous}}`, for example. See https://respec.org/docs/#data-link-for

JuhaVainio replied:
Removed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/85adbbee0d394d07645d3fab46cfe4c771b7f7e3

rakuco commented on index.html:
```diff
@@ -688,11 +688,18 @@ <h4>
         The [=remote end steps=] are:
       </p>
       <ol class="algorithm" data-cite="WEBDRIVER2">
+        <li>Let |posture| be the result of invoking get a property "posture"
```
You need to reference "get a property" appropriately.

rakuco replied:
Furthermore, "get a property" will give you some JavaScript value that can technically be of any type (including undefined if it was not passed in the HTTP request). You also need to ensure it's a string. See e.g. step 2 of the remote end steps of https://w3c.github.io/webdriver/#dfn-dispatch-actions-for-a-string or step 2 of https://w3c.github.io/webdriver/#send-alert-text.

JuhaVainio replied:
Would you know how to make the link? It should point to this https://w3c.github.io/webdriver/#dfn-getting-properties

In case of https://www.w3.org/TR/webdriver2/#dfn-extension-commands, `[=extension commands=]` works.

But in this case `[=getting properties=]` doesn't work. It is defined in here https://github.com/w3c/webdriver/blob/master/index.html#L362C18-L363C36
```
<dfn data-lt="getting properties|getting the
 property">getting a property</dfn>

```
I have tried also 
[=getting the property=]
[=getting a property=]

rakuco replied:
The problem's that the `<dfn>` is not exported by the WebDriver spec. This has also [affected the Generic Sensor spec](https://github.com/w3c/sensors/blob/d8b0f67c3acc27330f97554185a32577a7d9d063/index.bs#L55).

In ReSpec, the easiest workaround is to use the [`data-cite` attribute](https://respec.org/docs/#data-cite).

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/6bca850082316fb3e17953e1a95888300cef6569

JuhaVainio replied:
https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/0561e510aeaa4711062f53c7c45ddaa5511f6e7a added to fulfill the "need to ensure it's a string" requirement

rakuco commented on index.html:
```diff
@@ -688,8 +688,9 @@ <h4>
       <ol class="algorithm" data-cite="WEBDRIVER2">
-        <li>If |posture| is not {{DevicePostureType}}, return [=error=] with
-            [=error code|WebDriver error code=] [=invalid argument=].
+        <li>If |posture| is neither <a>continuous</a> nor <a>folded</a>, return
```
nit: since these are strings, they should be in quotes (e.g. `"<a>continuous</a>"` or `{{DevicePosture/continuous}}`).

JuhaVainio replied:
Hmmm... I hope you really meant 
{{DevicePosture/continuous}} 
and not actually 
"{{DevicePosture/continuous}}"
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/85adbbee0d394d07645d3fab46cfe4c771b7f7e3

rakuco commented on index.html:
```diff
@@ -688,11 +688,18 @@ <h4>
-            [=invalid argument=].
+        [=error=] with [=error code|WebDriver error code=] [=invalid argument=].
+        </li>
+        <li>Let |document| be the |context|'s [=navigable/active document=].
```
Where's `context` coming from?

JuhaVainio replied:
It came from earlier version. But it seems that you modified the document 
https://github.com/w3c/device-posture/commit/d88a411a8a9ca065719770d2c59eed617629234c#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051L479
So this line and other lines need to be checked against new latest version.

JuhaVainio replied:
Seems that in new version "document" is defined in this sentence:
"_The device posture change steps for a [Document](https://dom.spec.whatwg.org/#document) document and an optional boolean disallowRecursion (default false) are as follows:_"

So automation chapter need these same steps? Or why WebDriver cannot reference these steps like you have:
"`Run the [=device posture change steps=] on |document|.`" ?

rakuco replied:
The variable name may have changed, but the problem is the same in both versions: `|document|` and `|context|` were defined in a different section as part of a different algorithm, you cannot just reference them in your WebDriver endpoint algorithm as if this algorithm is running as part of the others. It's like defining a function `foo()` in C and referencing a local variable set in another function `bar()`.

You need to obtain the variables you need from _somewhere_. See the endpoints defined in the Generic Sensor spec, they also need to get a top-level traversable somehow (and from a top-level traversable, which is a navigable, you can basically obtain everything else you might need).

JuhaVainio replied:
I was asking: do I (can I) reference (meaning there would be less steps in Automation section) the **algorithm** (https://www.w3.org/TR/device-posture/#dfn-device-posture-change-steps) or do I write same type of steps to Automation chapter?
I was not asking can I say "_variables are defined same way as in #dfn-device-posture-change-steps_"

rakuco replied:
> I was asking: do I (can I) reference (meaning there would be less steps in Automation section) the **algorithm** (https://www.w3.org/TR/device-posture/#dfn-device-posture-change-steps)

Yes, doing that is what I've been saying (and helping prepare for) all along. If you don't do it, you'll be in the same situation as before, where you duplicate the existing algorithms instead of integrating with them.

JuhaVainio replied:
Rebasing somehow lost the location of this comment from Files-section. 

I added https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/8776797da5528953d6dd20edcca760fa74447b6d which starts to use newly defined "device posture change steps" algorithm. Is seems to need "document" parameter. In updated "8.2 Device Posture change" there is line "_it MUST run the device posture change steps with the top-level traversable's active document._". I would think WebDriver would need to do same and use same terms: "top-level traversable's" & "active document". Which I updated to spec. But now the `context` is not missing and I do not know it that needs to be fitted somewhere.

rakuco commented on index.html:
```diff
@@ -688,11 +688,18 @@ <h4>
+        </li>
+        <li>Let |document| be the |context|'s [=navigable/active document=].
+        </li>
+        <li><a>Update the device posture information</a> of |document|.
```
You should settle on one style: either `[=foo=]` or `<a>foo</a>`, as using both for the same purpose in the same algorithm is confusing.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/0148fe38e98f8870e7f7515fafcc1b2743641497

rakuco reviewed (changes requested):
As I mentioned in this week's call, the current version of the WebDriver endpoint is duplicating a lot of the behavior of the existing algorithms instead of working _with_ them:
- You are potentially changing the value of the `[[CurrentPosture]]` internal slot even when a document's visibility is hidden.
- There is no description of how invoking "update the device posture information" is going to take the value passed to WebDriver into consideration.

Furthermore, you are using the `context` variable without defining it anywhere, which is directly related to another issue: should a change like this affect a single document or all the documents (and frames) in a tab? Real posture changes behave like the latter, and it does not make sense to me to have an iframe report a different posture than its parent with WebDriver.

We also discussed how the existing algorithms were a bit vague and made it harder to integrate with them. I've attempted to improve the situation upstream with w3c/device-posture#120. If you use this pull request, you could:
- Define some DevicePostureType override associated with a [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversables) (take some time to read at least the beginning of https://html.spec.whatwg.org/multipage/document-sequences.html#infrastructure-for-sequences-of-documents). See https://w3c.github.io/sensors/#virtual-sensors for inspiration.
- Make the WebDriver command set this override to the value it receives, and add a way to set it to null as well so that it can be cleared.
- In the "calculate the device posture information" algorithm, add a step that checks if a document's navigable's top-level traversable has the override above set, and use it instead when it is.

JuhaVainio committed:
Remove data-link-for="DevicePostureType" from section tag

After removing data-link-for from section tag, links to
DevicePostureType had to reformatted.

rakuco commented on index.html:
```diff
@@ -618,6 +618,94 @@ <h4>
+        <li>Let |posture| be the result of invoking get a property "posture"
+        from |parameters|.
+        </li>
+        <li>If |posture| is neither {{DevicePostureType/continuous}} nor
+        {{DevicePostureType/folded}}, return [=error=] with
+        [=error code|WebDriver error code=] [=invalid argument=].
+        </li>
```
You asked somewhere if I meant `{{foo/bar}}` or `"{{foo/bar}}"` here, and it's the latter. You are referencing strings, so you should follow https://infra.spec.whatwg.org/#strings (note that a lot of specs do not follow the monospace font recommendation, the double quotes part is more important). See all the strings referenced in https://w3c.github.io/webdriver/#print-page for example.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/256310eb663523fe4e7ba3c10c777f429310c395

JuhaVainio committed:
Link "get a property" to #dfn-getting-properties in WebDriver specification

JuhaVainio committed:
Add quotation marks around continuous- and folded-strings

JuhaVainio committed:
Unify tag style

JuhaVainio committed:
Check posture is string

JuhaVainio committed:
Use new "device posture change steps" algorithm in "Set device posture"

"device posture change steps" has document-parameters, which was defined
in "Set device posture" steps.

JuhaVainio committed:
Clarify that document is argument to "device posture change steps" algorithm

rakuco commented on index.html:
```diff
@@ -614,6 +614,97 @@ <h4>
+        "{{DevicePostureType/folded}}", return [=error=] with
+        [=error code|WebDriver error code=] [=invalid argument=].
+        </li>
+        <li>Let |document| be the [=top-level traversable=]'s
```
We're almost there, but now the question is where is your top-level traversable coming from :-) You're referring to a general concept in the HTML spec, not something that was derived from something else (which would be the case if you had referred to https://html.spec.whatwg.org/multipage/document-sequences.html#nav-top instead, for example).

As I said, look at the Generic Sensors spec, it needed to deal with this very same problem.

JuhaVainio replied:
Please clarify. This is one example of Generic sensor spec:
"Let topLevelVirtualSensorMapping be the [current browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)'s [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#bc-traversable)'s [virtual sensor mapping](https://www.w3.org/TR/generic-sensor/#virtual-sensor-mapping)."

And this is device-posture:
"Let document be the [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable)'s [active document](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-document)."

Is the difference "[current browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)'s"?  So it should say:
"Let document be the [current browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)'s [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable)'s [active document](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-document)."?

rakuco replied:
> Is the difference "[current browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)'s"? So it should say:
> "Let document be the [current browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)'s [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable)'s [active document](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-document)."?

Yes (as explained in our call), but not exactly: your internal slot should be tied to a top-level traversable, not a Document.

JuhaVainio replied:
Fixed https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/a9d3d5a7bfc454eb439350719c23b4a0c1935c74

rakuco commented on index.html:
```diff
@@ -614,6 +614,97 @@ <h4>
+        <li>Let |document| be the [=top-level traversable=]'s
+        [=navigable/active document=].
+        </li>
+        <li>Invoke [=device posture change steps=] with |document| as its
```
You don't need the "as its argument" part because it's redundant. See the suggested forms in https://infra.spec.whatwg.org/#algorithm-params.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/6cd7b18eb6157ea0b2d887830360ee7a6029b85f

rakuco reviewed (changes requested):
This is better, but you're not doing anything with `|posture|`, so any value that is passed to this endpoint is not having any effect anywhere -- your invocation of "device posture change steps" is just running the regular algorithm as if it had been called due to a change in real posture.

See my suggestion in the previous comment again: the idea is to have an internal slot in the top-level traversable that overrides the posture and is read by https://w3c.github.io/device-posture/#calculating-the-device-posture-information.

Like I said before, this is a variation on something that the Generic Sensor spec already had to deal with (the only difference here is that you also need a way to disable the override, either by allowing a special value here or adding an endpoint that listens to a DELETE http request instead of POST).

JuhaVainio commented:
> Like I said before, this is a variation on something that the Generic Sensor spec already had to deal with (the only difference here is that you also need a way to disable the override, either by allowing a special value here or adding an endpoint that listens to a DELETE http request instead of POST).

Could you clarify the overall architecture you define.
So the current set_device_posture is not enough, we need more services? And what it currently does is not enough?
So when set_device_posture reaches chrome/test/chromedriver/window_commands.cc it should somehow read and save current posture and then modifies it with setDeviceMetricsOverride(). Then we have other service clear_device_posture which reads this saved value from somewhere ("slot in top-level-traversable") I writes it using setDeviceMetricsOverride()?

> internal slot in the top-level traversable

This went totally over my head. Is "Each [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable) has a virtual sensor mapping, which is an [ordered map](https://infra.spec.whatwg.org/#ordered-map) of [virtual sensor types](https://www.w3.org/TR/generic-sensor/#virtual-sensor-type) to [virtual sensor](https://www.w3.org/TR/generic-sensor/#virtual-sensor)." a slot in top-level traversable?

JuhaVainio committed:
Remove "as its argument"

JuhaVainio commented on index.html:
```diff
@@ -614,6 +614,96 @@ <h4>
     </section>
+    <section>
+      <h2>
+        Automation
```
I add my comment here as I do not see how your reply in Conversation page can be linked to my initial question and keep together.

> Like I said before, this is a variation on something that the Generic Sensor spec already had to deal with (the only difference here is that you also need a way to disable the override, either by allowing a special value here or adding an endpoint that listens to a DELETE http request instead of POST).

Could you clarify the overall architecture you define.
So the current set_device_posture is not enough, we need more services? And what it currently does is not enough?
So when set_device_posture reaches chrome/test/chromedriver/window_commands.cc it should somehow read and save current posture and then modifies it with setDeviceMetricsOverride(). Then we have other service clear_device_posture which reads this saved value from somewhere ("slot in top-level-traversable") I writes it using setDeviceMetricsOverride()?

> internal slot in the top-level traversable

This went totally over my head. Is "Each [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable) has a virtual sensor mapping, which is an [ordered map](https://infra.spec.whatwg.org/#ordered-map) of [virtual sensor types](https://www.w3.org/TR/generic-sensor/#virtual-sensor-type) to [virtual sensor](https://www.w3.org/TR/generic-sensor/#virtual-sensor)." a slot in top-level traversable? If it is where in sensor code this "slot" is handled/accessed? It try to understand how it would affect current (draft) Device Posture WebDriver code.

JuhaVainio replied:
I assume this is 99% wrong, but this has "clear device posture", some mods to "calculate the device posture information" and field in internal slot (assuming it is totally wrong location in spec)

https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/a9d3d5a7bfc454eb439350719c23b4a0c1935c74

JuhaVainio committed:
Add OverridePosture slot to document

Also Clear device posture is added to reprent OverridePosture usage.

rakuco commented on index.html:
```diff
@@ -182,6 +182,14 @@ <h3>
           </tr>
+          <tr>
+            <td>
+              <dfn data-dfn-for="Document">[[\OverridePosture]]</dfn>
```
* "PostureOverride" sounds better than "OverridePosture".
* As discussed in our call, it makes more sense to associate this with a top-level traversable than a Document, as the posture needs to be the same for all navigables in a tab and you're only setting the internal slot for the top-level traversable's active document anyway.

JuhaVainio replied:
- "OverridePosture" => "PostureOverride" done in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/f4ec72af9c58a85392930d7d73ef006176c1194c
- Document => top-level traversable : Something like this https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/9a3c613454ea244a392247beec1ea602fbc8486f ?

rakuco commented on index.html:
```diff
@@ -182,6 +182,14 @@ <h3>
+              <dfn data-dfn-for="Document">[[\OverridePosture]]</dfn>
+            </td>
+            <td>
+              Posture to override posture defined by hardware. Used in testing.
```
```suggestion
              Posture to override posture defined by hardware. Used in automation.
```

rakuco replied:
I'd mention the possible values it can have: `"[=continuous=]", "[=folded=]" or null`.

JuhaVainio replied:
Ooops. Found this after pushing https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/9a3c613454ea244a392247beec1ea602fbc8486f. If spec doesn't need to explain null then maybe this sentence style is better than bullets (my bullets should be surrounded with `<ul>`, otherwise I get style error). Or it could also have "or `null`. Set value as `null` to defined posture by hardware."

JuhaVainio replied:
testing => automation fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/daf7abfa604cb53e4245f10209d9ca499b1eae8d

rakuco commented on index.html:
```diff
@@ -436,6 +444,11 @@ <h2>
           {{Document}} |document:Document| are as follows:
         </p>
         <ol class="algorithm">
+          <li>If |document:Document|.{{Document/[[OverridePosture]]}} contains
```
As discussed in our call, the internal slot should be associated with a top-level traversable. You also need a way to go from a Document to a top-level traversable.
* A Document has a _relevant global object_, which is a Window in this case.
* A [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window-navigable) gives you a navigable.
* A navigable gives you a top-level traversable.

"document's relevant global object's navigable's top-level traversable" is a mouthful, but it works.

JuhaVainio replied:
`"head explode"`

When you said to look what some HTML item contains, before this explanation I would not found that three level deep path.

- Document: I would have went first to this location https://html.spec.whatwg.org/multipage/dom.html#document. And tried to find something called "top-level traversable" or similar between `partial interface Document {` and `}`. And though I have to find it in there.  After I read "A [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window-navigable) gives you a navigable." I understood that when you said "containing" I would need to look after the `partial interface Document {` stuff. Even now if I look [Document](https://html.spec.whatwg.org/multipage/dom.html#document) page I do not know where you pulled `relevant global object` from that page.
- "A [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window-navigable) gives you a navigable.": **Maybe** after this I remember , when you would give me link to [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window) I would find what it contains. 
- "A navigable gives you a top-level traversable.": Not easy but to conclude this maybe I should go to [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) and understand that ... sorry lost... In [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) page does this "A parent, a [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) or null." line indicate if it is "top-level traversable"? Is there some field in document where the value is "A parent, a [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) or null."?


In https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cf6fcf35101ce123b86a527595b51c384627c744 I wrote this change

rakuco replied:
> `"head explode"`

That's where reading other specs is really helpful, as these concepts are in different parts of the HTML spec so one of the quickest ways to see what can be done is to see what others have already managed to do (getting a navigable from a document is very common, and you can use the Webdex page I sent you yesterday or even GitHub search for this).

> * Document: I would have went first to this location https://html.spec.whatwg.org/multipage/dom.html#document. And tried to find something called "top-level traversable" or similar between `partial interface Document {` and `}`. And though I have to find it in there.

Yeah, the concepts are in different parts of the spec. https://html.spec.whatwg.org/multipage/document-sequences.html contains several other algorithms related to Documents which are relevant to this discussion.

> After I read "A [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window-navigable) gives you a navigable." I understood that when you said "containing" I would need to look after the `partial interface Document {` stuff. Even now if I look [Document](https://html.spec.whatwg.org/multipage/dom.html#document) page I do not know where you pulled `relevant global object` from that page.

This one comes from https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global. "Global object" is a term that comes from JavaScript -- https://developer.mozilla.org/en-US/docs/Glossary/Global_object. The important thing to know in this case is `Window` is a kind of global object, and since this API is only exposed to the Window context (and not workers), a Document's relevant global object will always be a Window (and not a worker).

Finding all of this out on your own is very time-consuming, but reading what others have done helps **a lot** here (a lot of specs need to refer to "relevant global object").

> * "A navigable gives you a top-level traversable.": Not easy but to conclude this maybe I should go to [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) and understand that ... sorry lost... In [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) page does this "A parent, a [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) or null." line indicate if it is "top-level traversable"? Is there some field in document where the value is "A parent, a [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) or null."?

I mentioned this in another comment; the relevant definition here is https://html.spec.whatwg.org/multipage/document-sequences.html#nav-top, which is the definition of "a navigable's top-level traversable".

rakuco commented on index.html:
```diff
@@ -436,6 +444,11 @@ <h2>
         </p>
         <ol class="algorithm">
+          <li>If |document:Document|.{{Document/[[OverridePosture]]}} contains
+          "{{DevicePostureType/continuous}}" or "{{DevicePostureType/folded}}",
```
Instead of checking for specific values here, I'd just say the equivalent of "if it's not null, return it" (i.e. instead of checking for the presence of specific values, check for the absence of value).

JuhaVainio replied:
Sorry I fixed it same time as above in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cf6fcf35101ce123b86a527595b51c384627c744 , because it is easier to try to formulate the whole sentence than only part of it.

JuhaVainio committed:
Define internal slot for device posture to top-level traversable

Above change affected also "calculate the device posture information"
algorithm.

Set device posture service did not accurately defined which top-level
traversable it was accessing.

JuhaVainio committed:
Rephrase how to get to top-level traversable's OverridePosture

Also rephrase what kind of values to check from OverridePosture.

JuhaVainio committed:
Use word automation not testing

JuhaVainio committed:
Rename OverridePosture as PostureOverride

JuhaVainio committed:
Define Clear device posture service

Clear device posture is now described in much more detail.

rakuco commented on index.html:
```diff
@@ -184,6 +187,42 @@ <h3>
+      </h4>
+      <p>
+        The following internal slots are added to the
+        [=top-level traversable=]'s {{Document}} interface.
```
What we discussed yesterday was that this internal slot belongs to the top-level traversable itself, not to its Document. The previous version of your text was already setting this value only in the top-level traversable's Document, and my point was that instead of using the internal slot only in certain Documents, you should skip the Document part altogether and add this data to the top-level traversable.

Think about it this way too: if you set it to a Document, it will be unset whenever there's a navigation (i.e. you go from one page to another in the tab). If you assign it to the top-level traversable, the override will remain across navigations, which is the behavior we want.

JuhaVainio replied:
Sorry I thought the "document" in here is not other item but place/structure to have the internal slot. (Badly explained)

Next problem is word "interface" . That came from {{Document}}. Word "interface" cannot be used with "top-level traversable" ? These sentences come so weird. So is the header then
"[relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global)'s [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable)'s [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable)" OR just `[=Top-level traversable=]`?

rakuco replied:
"Interface" has a very specific meaning: https://webidl.spec.whatwg.org/#idl-interfaces. `Document` is an interface because it is declared as such in https://dom.spec.whatwg.org/#interface-document and further extended in the HTML spec.

A top-level traversable does not have a Web IDL definition and cannot be called an interface.

> So is the header then
"[relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global)'s [navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable)'s [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable)" OR just [=Top-level traversable=]?

The former. If you just say "the top-level traversable", you need to explain which top-level traversable you are referring to and where you are getting it from, which leads you back to the first version.

Another way to look at it is that "[=top-level traversable=]" is https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable, which doesn't tell you much in the context of the algorithm, and the other is https://html.spec.whatwg.org/multipage/document-sequences.html#nav-top.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/d9d848ca4807f9ca82e392f6305b9631daa6272f

JuhaVainio committed:
Move internal slot to top-level traversable

Earlier internal slot was in top-level traversable's Document. Now it is
moved to top-level traversable itself.

Also highlighted that it is the relevant global object's navigable's
top-level traversable and not just any top-level traversable.

rakuco commented on index.html:
```diff
@@ -184,6 +187,42 @@ <h3>
+            <td>
+              Posture to override posture defined by hardware. Used in
+              automation. Possible values:
+              <li>"{{DevicePostureType/continuous}}"
```
This markup is incomplete, you need a parent list element: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li

JuhaVainio replied:
Added in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/3f51f595e803619a4190efbcc400f2ff0ba66306

rakuco commented on index.html:
```diff
@@ -184,6 +187,42 @@ <h3>
           </tr>
         </tbody>
       </table>
+      <h4>
```
I think I finally understood what you meant by "header" in a previous comment.

You need to use "relevant global object's blablabla" in the algorithms, not here. Your algorithms are still broken, and these terms only make sense in the context of the algorithms, where you are given a Document `|document|` as input and which you use to derive all these other terms. Just saying "relevant global object" without retrieving it from somewhere does not work either.

What you need in this specific section is simpler:
- Drop the "The Document interface" header
- Drop this header
- Gather everything in one section just like before. The difference is that you're going to have two tables. The first one is the one that's already present in the spec; the second one is prefaced by something like _"[=Top-level traversables=] must have the following internal slots:"_. In this case, you're really referring to the _concept_ of top-level traversables, not the top-level traversable that belongs to one specific navigable.

JuhaVainio replied:
"Internal slot" stuff is fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/a9c074e99df1f1ef01f77e44df63d9df5d5cf459

**"_Your algorithms are still broken_"**
- Currently text in algorithm says`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`
- "_You need to use "relevant global object's blablabla" in the algorithms_" So it has already "relevant global object"?
- "_Just saying "relevant global object" without retrieving it from somewhere does not work either._" I though the  `[=navigable=]'s [=top-level traversable=]` part is the "_somewhere_", the top-level traversable of given document.

rakuco replied:
> "Internal slot" stuff is fixed in [a9c074e](https://github.com/JuhaVainio/W3C-device-posture/commit/a9c074e99df1f1ef01f77e44df63d9df5d5cf459)
> 
> **"_Your algorithms are still broken_"**
> 
> * Currently text in algorithm says`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`

If you think in terms of code, what you need is this:

```c++
struct TopLevelTraversable {
  DevicePostureType* override;
};
```

but what you are doing is

```c++
struct TopLevelTraversable {
};

struct Document {
  DevicePostureType* override;
};

void CalculatePostureInformation(Document doc) {
  if (doc.window().navigable().topLevelTraversable().override) {
    // ...
  }
}
```

`[[PostureOverride]]` is a term that references the top-level traversable. It does not make sense to `data-dfn-for` it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.

Furthermore, while "calculate the device posture information" is almost correct, your WebDriver algorithms are not.

> * "_You need to use "relevant global object's blablabla" in the algorithms_" So it has already "relevant global object"?

What do you mean by "it" in this context?

> * "_Just saying "relevant global object" without retrieving it from somewhere does not work either._" I though the  `[=navigable=]'s [=top-level traversable=]` part is the "_somewhere_", the top-level traversable of given document.

You retrieve the navigable from the relevant global object in your algorithms, not the other way around. In any case, what I'm saying is that in this location where you're defining the slot, you don't have a "given document" like in the algorithms where you're receiving a specific Document instance.

JuhaVainio replied:
The code for "8.1 Calculating the device posture information" is
`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`
the output text is
"_If document's relevant global object's navigable's top-level traversable.[[PostureOverride]] is non-null, return it._"
Clarification: Is the outputted text OK? So only the code is incorrect?

JuhaVainio replied:
> [[PostureOverride]] is a term that references the top-level traversable. It does not make sense to data-dfn-for it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.

When defining PostureOverride as internal slot `[[\InternalSlot]]` style is used.
If I use `<dfn>[[\InternalSlot]]</dfn>` browser says "Internal slot _name of slot_ must be associated with WebIDL interface". In earlier comment you said "top-level traversable does not have a Web IDL definition", so which interface it should be linked? On the other hand I noticed that I do not get the error if I write something there. For example
`<dfn data-dfn-for="TopLevelTraversable">[[\PostureOverride]]</dfn>`
This is little bit weird area as the text (_TopLevelTraversable_) in not shown in spec which you actually read and it doesn't affect the created definition/link to that definition.

JuhaVainio replied:
>>> "You need to use "relevant global object's blablabla" in the algorithms, not here."
>>
>>So it has already "relevant global object"?
>
> What do you mean by "it" in this context?

Flow of text + my assumption
1. Your text "_You need to use "relevant global object's blablabla" in the algorithms, not here._"
2. I assumed "_need to use_" is negative term and is meant for algorithm section. Meaning text in algorithm section does not fulfill this requirement.
3. Then I would go to "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" to search if "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" already uses term "_relevant global object's_". As "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" is already using "_relevant global object_" term I said said **it** (meaning "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)") has already "_relevant global object's_"

rakuco replied:
> The code for "8.1 Calculating the device posture information" is `If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null` the output text is "_If document's relevant global object's navigable's top-level traversable.[[PostureOverride]] is non-null, return it._" Clarification: Is the outputted text OK? So only the code is incorrect?

It's unusual to use the "foo.bar" notation when "foo" is not a variable itself. A more common way of writing this would be "document's relevant global object's navigable's top-level traversable's [[PostureOverride]]". Alternatively, you can break it down into more than one step. For example:

1. Let `|topLevelTraversable|` be `|document|`'s relevant global object's navigable's top-level traversable.
2. If `|topLevelTraversable|.[[PostureOverride]]` is [...]`.

> > [[PostureOverride]] is a term that references the top-level traversable. It does not make sense to data-dfn-for it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.
> 
> When defining PostureOverride as internal slot `[[\InternalSlot]]` style is used. If I use `<dfn>[[\InternalSlot]]</dfn>` browser says "Internal slot _name of slot_ must be associated with WebIDL interface". In earlier comment you said "top-level traversable does not have a Web IDL definition", so which interface it should be linked? On the other hand I noticed that I do not get the error if I write something there. For example `<dfn data-dfn-for="TopLevelTraversable">[[\PostureOverride]]</dfn>` This is little bit weird area as the text (_TopLevelTraversable_) in not shown in spec which you actually read and it doesn't affect the created definition/link to that definition.

The warning is a bit misleading, as you don't need to specify a Web IDL interface necessarily; any `<dfn>` works. I couldn't get `[[]]` to work with the `[= =]` shorthand syntax though; what did work is something like this: `<a data-link-for="top-level traversable">[[\PostureOverride]]</a>`.

JuhaVainio replied:
https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49

> any `<dfn>` works.

I assume you meant "any made up interface name". So I used "top-level traversable" in "data-dfn-for" when defining PostureOverride internal slot.

> 1. Let |topLevelTraversable|
> 2. If |topLevelTraversable|

In https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49 I have this two step approach.

rakuco replied:
> [cd4e35a](https://github.com/JuhaVainio/W3C-device-posture/commit/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49)
> 
> > any `<dfn>` works.
> 
> I assume you meant "any made up interface name". So I used "top-level traversable" in "data-dfn-for" when defining PostureOverride internal slot.

Well using it with a definition that does not exist does not make much sense. "top-level traversable" is a valid `<dfn>` coming from the HTML spec.

JuhaVainio committed:
Add parent element to bullet points

JuhaVainio committed:
Refer "top-level traversable" generally in "Internal slots" section not to specific "top-level traversable"

JuhaVainio committed:
Remove PostureOverride referring to Document.

rakuco commented on index.html:
```diff
@@ -737,7 +740,8 @@ <h4>
         context</a>'s [=top-level traversable=]'s [=navigable/active document=].
         </li>
-        <li>Set |document|.{{Document/[[PostureOverride]]}} to |posture|.
+        <li>Set |document|.<a data-link-for="top-level traversable">
```
This is still wrong.

JuhaVainio replied:
https://github.com/JuhaVainio/W3C-device-posture/commit/9192ea0141132ca88512739cf0d0d7db8e39dd40

rakuco commented on index.html:
```diff
@@ -777,7 +781,8 @@ <h4>
         context</a>'s [=top-level traversable=]'s [=navigable/active document=].
         </li>
-        <li>Set |document|.{{Document/[[PostureOverride]]}} to
+        <li>Set |document|.<a data-link-for="top-level traversable">
```
This is still wrong.

JuhaVainio replied:
Could you verbose a little bit. Which part is wrong? Code or the HTML it produces?
|document| in here doesn't mean the "Document" which every e.g. iframe has, but is variable of current browsing context's top-level traversable's active document. So should not be the `|document|` part. There should not be `.`  between `|document|` and `<a data-link-for` ?

rakuco replied:
We've been talking specifically about this for days so I thought I didn't need to elaborate further. If you say that `[[PostureOverride]]` is an internal slot of the top-level traversable, then accessing it from a Document is wrong.

JuhaVainio replied:
So the problems starts from already earlier step?

**Text**
"
Let _document_ be the current browsing context's top-level traversable's active document.
Set _document_.[[PostureOverride]] to _null_.
"
==>
"
Let _topLevelTraversable_ be the current browsing context's top-level traversable.
Set _topLevelTraversable_.[[PostureOverride]] to _null_.
"

**Code**
```
<li>Let |document| be the
<a data-cite="webdriver2/#dfn-current-browsing-context">current browsing
context</a>'s [=top-level traversable=]'s [=navigable/active document=].
</li>
<li>Set |document|.<a data-link-for="top-level traversable">
[[\PostureOverride]]</a> to <code>null</code>.
```
==>
```
<li>Let |topLevelTraversable| be the
<a data-cite="webdriver2/#dfn-current-browsing-context">current browsing
context</a>'s [=top-level traversable=].
</li>
<li>Set |topLevelTraversable|.<a data-link-for="top-level traversable">
[[\PostureOverride]]</a> to <code>null</code>.
```

rakuco replied:
> So the problems starts from already earlier step?

Yes.

Your proposed fix is almost correct, `[=top-level traversable=]` will refer to the concept, not the browsing context's top-level traversable, which you get with `[=browsing context/top-level traversable=]`. Again, this is all in the Generic Sensor spec.

JuhaVainio replied:
https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/9192ea0141132ca88512739cf0d0d7db8e39dd40

Damn there was still that.

JuhaVainio committed:
In WebDriver steps do not use Document but top-level traversable

rakuco commented on index.html:
```diff
@@ -614,6 +655,142 @@ <h4>
+        </tr>
+      </table>
+      <p>
+        This [=extension commands|extension command=] changes device posture to a specific
```
```suggestion
        This [=extension command=] changes device posture to a specific
```

Respec handles singular and plural nouns automatically.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/70bba088084478e20882f7773dbd275b3e5cc891

rakuco commented on index.html:
```diff
@@ -614,6 +655,142 @@ <h4>
+        <a data-cite="webdriver2/#dfn-current-browsing-context">current browsing
+        context</a>'s [=browsing context/top-level traversable=].
+        </li>
+        <li>Set |topLevelTraversable|.<a data-link-for="top-level traversable">
```
I think before this step you should also check if `[[PostureOverride]]` is null and return if so, otherwise this command will always trigger the _device posture change steps_.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/adeab1c64d6b7198a23297c8963f9b0fc2bc2ee4

What about in Set device posture?
```
4. Let topLevelTraversable be the current browsing context's top-level traversable.
5. Set topLevelTraversable.[[PostureOverride]] to |posture|.
```
==>
```
4. Let topLevelTraversable be the current browsing context's top-level traversable.
5. If topLevelTraversable.[[PostureOverride]] is equal to |posture|, return success with data null.
6. Set topLevelTraversable.[[PostureOverride]] to posture.
```

rakuco replied:
A similar check already happens in step 3 of https://w3c.github.io/device-posture/#dfn-device-posture-change-steps.

The problem with clearing the posture override specifically was that when an override was not set it was a sure way of triggering the non-override (i.e. platform- and OS-dependent) calculations.

rakuco commented on index.html:
```diff
@@ -614,6 +655,142 @@ <h4>
+        [[\PostureOverride]]</a> to
+        <code>null</code>.
+        </li>
+        <li>Invoke [=device posture change steps=] with |document|.
```
`|document|` is not defined. You have to derive it from the top-level traversable (`|document|` is its active document, as you might remember from previous versions of this text).

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/9ede2a7cd27920136bd55036225e00e16c5be77d

rakuco commented on index.html:
```diff
@@ -614,6 +655,142 @@ <h4>
+        <li>Set |topLevelTraversable|.<a data-link-for="top-level traversable">
+        [[\PostureOverride]]</a> to |posture|.
+        </li>
+        <li>Invoke [=device posture change steps=] with |document|.
```
`|document|` is not defined. You have to derive it from the top-level traversable (`|document|` is its active document, as you might remember from previous versions of this text).

JuhaVainio replied:
Can I use |topLevelTraversable| variable?
`Let |document| be |topLevelTraversable|'s [=navigable/active document=]`
so it doesn't need to be
`Let |document| be the  <a data-cite="webdriver2/#dfn-current-browsing-context">current browsing context</a>'s [=browsing context/top-level traversable=]'s [=navigable/active document=].`

rakuco replied:
Sure. They just different ways of referring to the same thing.

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/9ede2a7cd27920136bd55036225e00e16c5be77d

rakuco reviewed (changes requested):
It's almost there. I think once the new inline comments are addressed it will be ready for shipping.

JuhaVainio committed:
Let ReSpec take care of singular and plural nouns automatically

JuhaVainio committed:
In Clear device posture check if [[PostureOverride]] is null and return if so

JuhaVainio committed:
Convert plain success texts to links to WebDriver success return value

JuhaVainio committed:
Define where document-variable is derived from

rakuco commented on index.html:
```diff
@@ -614,6 +655,152 @@ <h4>
+        <code>null</code>.
+        </li>
+        <li>Let |document| be |topLevelTraversable|'s [=navigable/active
+        document=]
```
```suggestion
        document=].
```

JuhaVainio replied:
Remembered in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/facde85c733cf1b19ad25a6eef6a281eed7e3fa9

rakuco commented on index.html:
```diff
@@ -614,6 +655,152 @@ <h4>
+        [[\PostureOverride]]</a> to |posture|.
+        </li>
+        <li>Let |document| be |topLevelTraversable|'s [=navigable/active
+        document=]
```
```suggestion
        document=].
```

JuhaVainio replied:
Fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/facde85c733cf1b19ad25a6eef6a281eed7e3fa9

rakuco reviewed (approved):
lgtm with a few nits

JuhaVainio committed:
Add missing full stops to end of sentences.
